integration_tests/src/main/python/join_test.py

```diff
@@ -1026,3 +1026,68 @@ def do_join(spark):
         right = spark.read.parquet(data_path2)
         return right.filter("cast(id2 as bigint) % 3 = 4").join(left, left.id == right.id, "inner")
     assert_gpu_and_cpu_are_equal_collect(do_join, bloom_filter_confs)
+
+
+@ignore_order(local=True)
+@allow_non_gpu('ProjectExec', 'BroadcastHashJoinExec', 'BroadcastExchangeExec',
+               'FilterExec', 'ColumnarToRowExec')
+@pytest.mark.parametrize("is_gpu_parquet", [True, False],
+                         ids=["GpuParquetScan", "ParquetScan"])
+@pytest.mark.parametrize("is_gpu_broadcast", [True, False],
+                         ids=["GpuBroadcastExchange", "BroadcastExchange"])
+def test_broadcast_hash_join_fix_fallback_by_inputfile(spark_tmp_path, is_gpu_parquet,
```

Comment:
I ran these tests on the current 23.12 and `test_broadcast_hash_join_fix_fallback_by_inputfile[BroadcastExchange-ParquetScan] ` produced the wrong answer, but `test_broadcast_hash_join_fix_fallback_by_inputfile[GpuBroadcastExchange-ParquetScan]` failed with not falling back as expected.

```
java.lang.AssertionError: assertion failed: Could not find BroadcastHashJoinExec in the Spark plan
```

`test_broadcast_nested_join_fix_fallback_by_inputfile` passed in all cases and none of them triggered the error as described in https://github.com/NVIDIA/spark-rapids/issues/9469

Can we please add in a test that is the same as #9469 so we can be sure that it is fixed?

Reply:
The case in https://github.com/NVIDIA/spark-rapids/issues/9469 requires Iceberg to run, so we can not test this for Spark 330+, is it OK?
And can you share more about how you run this IT test, then I can try to repro the errors to see if I am missing anything to make it always pass.

Reply:
I updated the tests, now they can reproduce the same error as https://github.com/NVIDIA/spark-rapids/issues/9469 on the current 23.12.
```
E                               raise Py4JJavaError(
                        "An error occurred while calling {0}{1}{2}.\n".
>                       format(target_id, ".", name), value)
E                   py4j.protocol.Py4JJavaError: An error occurred while calling o659.collectToPython.
E                   : java.lang.IllegalStateException: the broadcast must be on the GPU too
E                   	at com.nvidia.spark.rapids.shims.GpuBroadcastJoinMeta.verifyBuildSideWasReplaced(GpuBroadcastJoinMeta.scala:72)
E                   	at org.apache.spark.sql.rapids.execution.GpuBroadcastNestedLoopJoinMeta.convertToGpu(GpuBroadcastNestedLoopJoinExec.scala:59)
E                   	at org.apache.spark.sql.rapids.execution.GpuBroadcastNestedLoopJoinMeta.convertToGpu(GpuBroadcastNestedLoopJoinExec.scala:45)
E                   	at com.nvidia.spark.rapids.SparkPlanMeta.convertIfNeeded(RapidsMeta.scala:799)





                    raise Py4JJavaError(
                        "An error occurred while calling {0}{1}{2}.\n".
>                       format(target_id, ".", name), value)
E                   py4j.protocol.Py4JJavaError: An error occurred while calling o663.collectToPython.
E                   : java.lang.IllegalStateException: the broadcast must be on the GPU too
E                   	at com.nvidia.spark.rapids.shims.GpuBroadcastJoinMeta.verifyBuildSideWasReplaced(GpuBroadcastJoinMeta.scala:72)
E                   	at org.apache.spark.sql.rapids.execution.GpuBroadcastHashJoinMeta.convertToGpu(GpuBroadcastHashJoinExec.scala:63)
E                   	at org.apache.spark.sql.rapids.execution.GpuBroadcastHashJoinMeta.convertToGpu(GpuBroadcastHashJoinExec.scala:44)
E                   	at com.nvidia.spark.rapids.SparkPlanMeta.convertIfNeeded(RapidsMeta.scala:799)
```

Reply:
@revans2 Could you take a look again? Thx in advance.